File: web/static/modelspace.js
/* Model space: the migration slider, the colour year, and the what-if knobs.
 *
 * Deliberately a separate Alpine component from `detail` in app.js, with its own swatch and
 * its own canvas, rather than driving the page's hero. Two reasons: the panels then read as a
 * side-by-side "what if" against the real colour that stays on screen above them — which is
 * the honest framing, since none of this is the planet — and app.js is left untouched, which
 * is what keeps parallel sessions on this repo from colliding (see CLAUDE.md).
 *
 * Every colour shown here is precomputed in Python (web/modelspace.py); this file only picks
 * between stops and interpolates between neighbours. It never does colour science of its own.
 */
(function () {
  "use strict";

  function hexToRgb(h) {
    var s = String(h || "").replace("#", "");
    if (s.length !== 6) return [0, 0, 0];
    return [parseInt(s.slice(0, 2), 16), parseInt(s.slice(2, 4), 16), parseInt(s.slice(4, 6), 16)];
  }

  function two(n) {
    var s = Math.max(0, Math.min(255, Math.round(n))).toString(16);
    return s.length < 2 ? "0" + s : s;
  }

  /* Blend two track stops. Only ever used BETWEEN adjacent precomputed stops, which sit a
   * fraction of an octave apart, so a plain sRGB mix cannot wander anywhere the physics did
   * not already go — it just stops the swatch from stepping visibly as the slider moves. */
  function mixHex(a, b, t) {
    var x = hexToRgb(a), y = hexToRgb(b);
    return "#" + two(x[0] + (y[0] - x[0]) * t) + two(x[1] + (y[1] - x[1]) * t) +
      two(x[2] + (y[2] - x[2]) * t);
  }

  /* Distances span nine orders of magnitude across the archive; one format does not fit. */
  function fmtAu(au) {
    if (au == null) return "—";
    if (au < 0.01) return au.toFixed(4).replace(/0+$/, "").replace(/\.$/, "") + " AU";
    if (au < 1) return au.toFixed(3) + " AU";
    if (au < 100) return au.toFixed(2) + " AU";
    return Math.round(au).toLocaleString() + " AU";
  }

  function fmtTemp(k) {
    return Math.round(k).toLocaleString() + " K";
  }

  /* Plain-English regime name for a temperature. The slider's whole point is that the planet
   * crosses these, so naming them is what turns a moving swatch into an explanation. */
  function regime(k) {
    if (k >= 1600) return "ultra-hot — cloud-free, very dark";
    if (k >= 900) return "hot — clouds gone, sodium eats the yellow";
    if (k >= 500) return "warm — methane breaking up, hazy";
    if (k >= 220) return "temperate — water and ammonia clouds";
    return "cold — thick cloud, methane frozen in";
  }

  /* Each what-if variant's wording is identical on every planet page, so it is shipped once
   * here rather than inlined 5,800 times. Keys must match the ids in pipeline/modelspace.py. */
  var WHATIF_TEXT = {
    "clouds-off": {
      label: "No clouds at all",
      detail: "Strip the cloud deck and you see straight down into the deep atmosphere: much " +
        "darker, and bluer where Rayleigh scattering takes over.",
    },
    "clouds-thick": {
      label: "Solid cloud deck",
      detail: "Close the cloud deck completely. Clouds are bright and fairly grey, so this " +
        "pushes almost any planet toward a pale version of its star's own colour.",
    },
    "metal-poor": {
      label: "Solar metallicity (1×)",
      detail: "As few heavy elements as the Sun. Less methane to absorb the red end, so the " +
        "colour drifts back toward the star's.",
    },
    "metal-rich": {
      label: "Metal-rich (30×)",
      detail: "Thirty times the Sun's heavy elements — Neptune territory. Deeper methane " +
        "bands eat the red and the planet swings blue-green.",
    },
  };

  function textFor(id) {
    return WHATIF_TEXT[id] || { label: id, detail: "" };
  }

  document.addEventListener("alpine:init", function () {
    Alpine.data("modelspace", function (init) {
      return {
        // Parallel arrays: {r, au, t, h, l}, each of the same length. See web/modelspace.py.
        stops: init.stops || { r: [], au: [], t: [], h: [], l: [] },
        home: init.home || 0,
        cahoy: init.cahoy || [],
        // Per-planet numbers only; the wording is joined on from WHATIF_TEXT above.
        whatif: (init.whatif || []).map(function (v) {
          return {
            id: v.id, h: v.h, l: v.l, de: v.de,
            label: textFor(v.id).label, detail: textFor(v.id).detail,
          };
        }),
        year: init.year || null,
        measuredBase: !!init.measuredBase,
        radius: init.radius || 8,
        cloudState: init.cloudState || "",
        planetHex: init.planetHex || "#888888",

        idx: 0,          // slider position, integer index into stops
        frac: 0,         // fractional offset while the colour year is playing
        playing: false,
        info: "",        // which ℹ explainer is open ("move" | "year" | "whatif")
        picked: null,    // selected what-if variant id, null = the planet as modelled
        _raf: null,
        _t0: 0,
        _lastPaint: "",   // last hex actually drawn, so repaints of the same colour are free

        init() {
          this.idx = this.home;
          this.$nextTick(() => { this.paint(); this.paintWhatIf(); });
        },

        // --- the current position -------------------------------------------------------
        n() { return this.stops.h.length; },
        clamp(i) { return Math.max(0, Math.min(this.n() - 1, i)); },
        pos() { return this.idx + this.frac; },
        iLo() { return this.clamp(Math.floor(this.pos())); },
        iHi() { return this.clamp(Math.ceil(this.pos())); },
        t() { return this.pos() - Math.floor(this.pos()); },
        /* Linear blend of one parallel array between the two bracketing stops. */
        lerp(key) {
          var a = this.stops[key][this.iLo()], b = this.stops[key][this.iHi()];
          if (a == null) return null;
          if (b == null) b = a;
          return a + (b - a) * this.t();
        },

        hex() { return mixHex(this.stops.h[this.iLo()], this.stops.h[this.iHi()], this.t()); },
        lum() { var v = this.lerp("l"); return v == null ? 0 : v; },
        au() {
          var a = this.stops.au[this.iLo()];
          if (a == null) return null;
          // Distance is log-spaced, so interpolate it that way — a linear mix between two
          // stops an octave apart would print a distance the colour was never computed at.
          var b = this.stops.au[this.iHi()];
          if (b == null) b = a;
          return a <= 0 || b <= 0 ? a : Math.exp(Math.log(a) + (Math.log(b) - Math.log(a)) * this.t());
        },
        temp() { var v = this.lerp("t"); return v == null ? 0 : v; },
        atHome() { return Math.abs(this.pos() - this.home) < 0.001; },

        // --- readouts -------------------------------------------------------------------
        auText() { return fmtAu(this.au()); },
        tempText() { return fmtTemp(this.temp()); },
        regimeText() { return regime(this.temp()); },
        // "3.2× further out" / "12× closer in" — the comparison a newcomer actually wants.
        moveText() {
          var r = this.stops.r[this.iLo()];
          if (r == null) r = 1;
          if (this.atHome()) return "where it really is";
          var f = r >= 1 ? r : 1 / r;
          var n = f >= 100 ? Math.round(f).toLocaleString() : f.toFixed(f < 10 ? 1 : 0);
          return n + "× " + (r >= 1 ? "further out" : "closer in");
        },
        lumText() { return (this.lum() * 100).toFixed(1) + "%"; },

        // --- interactions ---------------------------------------------------------------
        moved() { this.stopYear(); this.frac = 0; this.paint(); },
        goHome() { this.stopYear(); this.idx = this.home; this.frac = 0; this.paint(); },
        toggleInfo(k) { this.info = this.info === k ? "" : k; },

        /* Jump the slider to a Cahoy grid distance, so the two models can be read against
         * each other at the same place rather than by eye across the control. */
        gotoCahoy(c) {
          var best = 0, bestD = Infinity;
          for (var i = 0; i < this.n(); i++) {
            var d = Math.abs(Math.log(this.stops.r[i]) - Math.log(c.r));
            if (d < bestD) { bestD = d; best = i; }
          }
          this.stopYear();
          this.idx = best; this.frac = 0;
          this.paint();
        },
        /* Where a grid mark sits along the slider, as a percentage of its width. */
        cahoyPct(c) {
          var lo = Math.log(this.stops.r[0]), hi = Math.log(this.stops.r[this.n() - 1]);
          return (100 * (Math.log(c.r) - lo) / (hi - lo)).toFixed(2) + "%";
        },
        homePct() {
          return (100 * this.home / Math.max(1, this.n() - 1)).toFixed(2) + "%";
        },

        // --- the colour year ------------------------------------------------------------
        /* One orbit in 12 s of wall clock, sampled at equal steps of TIME. The planet
         * therefore loiters through the cold outer stretch and whips through periastron,
         * which is the honest shape of an eccentric year and the thing worth watching. */
        toggleYear() { this.playing ? this.stopYear() : this.startYear(); },
        startYear() {
          if (!this.year || this.playing) return;
          this.playing = true;
          this._t0 = 0;
          var self = this;
          var lastFrame = 0;
          var step = function (now) {
            if (!self.playing) return;
            // Reschedule BEFORE doing any work: a throw or a slow frame in the repaint below
            // must never be able to leave the loop dead with `playing` still true.
            self._raf = requestAnimationFrame(step);
            // A WebGL repaint costs ~13 ms and the page hero is already spinning its own, so
            // running this at display rate starves the frame budget and stalls the tab. One
            // orbit takes 12 s; 20 fps is far more resolution than that motion needs.
            if (now - lastFrame < 50) return;
            lastFrame = now;
            if (!self._t0) self._t0 = now;
            var pos = self.year.pos;
            var phase = (((now - self._t0) / 12000) % 1) * pos.length;
            var i = Math.floor(phase), f = phase - i;
            var a = pos[i % pos.length], b = pos[(i + 1) % pos.length];
            var track = a + (b - a) * f;
            self.idx = Math.floor(track);
            self.frac = track - self.idx;
            self.paint();
          };
          this._raf = requestAnimationFrame(step);
        },
        stopYear() {
          this.playing = false;
          if (this._raf) { cancelAnimationFrame(this._raf); this._raf = null; }
        },
        /* Fraction of the orbit the planet spends near periastron, as plain English. */
        hotText() {
          if (!this.year) return "";
          var pct = this.year.hot * 100;
          return pct < 1 ? "under 1% of its year" : Math.round(pct) + "% of its year";
        },

        // --- what-if --------------------------------------------------------------------
        pick(id) { this.picked = this.picked === id ? null : id; this.paintWhatIf(); },
        current() {
          var self = this;
          return this.whatif.filter(function (v) { return v.id === self.picked; })[0] || null;
        },
        whatIfHex() { var v = this.current(); return v ? v.h : this.planetHex; },
        whatIfLabel() { var v = this.current(); return v ? v.label : "As modelled"; },
        whatIfDetail() {
          var v = this.current();
          return v ? v.detail
            : "The assumptions the pipeline actually used for this planet. Pick one of the " +
              "others to see how far the colour moves when that assumption changes.";
        },
        whatIfDe() { var v = this.current(); return v ? v.de : 0; },
        /* How to read a ΔE, in words rather than as a bare number. */
        deText() {
          var de = this.whatIfDe();
          if (!this.picked) return "this planet's own assumptions";
          if (de < 1) return "no visible difference";
          if (de < 2) return "barely visible";
          if (de < 10) return "clearly a different colour";
          return "a completely different colour";
        },

        // --- rendering ------------------------------------------------------------------
        /* A repaint costs ~13 ms, and both callers can fire far faster than that: a slider
         * drag emits input events at display rate, and the colour year ticks 20 times a
         * second. The colour itself only takes 25 distinct values across the whole track, so
         * skipping repaints that would draw the same disc again is nearly all of the work. */
        paint() {
          if (!window.PlanetRender || !this.$refs.msCanvas) return;
          var hex = this.hex();
          if (hex === this._lastPaint) return;
          this._lastPaint = hex;
          window.PlanetRender.render(this.$refs.msCanvas, {
            palette: window.PlanetRender.ramp(hex),
            baseHex: hex,
            radius: this.radius,
            cloudState: this.cloudState,
            lumY: this.lum(),
            style: "smooth",
          });
        },
        paintWhatIf() {
          if (!window.PlanetRender || !this.$refs.wiCanvas) return;
          var v = this.current();
          var hex = v ? v.h : this.planetHex;
          window.PlanetRender.render(this.$refs.wiCanvas, {
            palette: window.PlanetRender.ramp(hex),
            baseHex: hex,
            radius: this.radius,
            cloudState: this.cloudState,
            lumY: v ? v.l : null,
            style: "smooth",
          });
        },
        destroy() { this.stopYear(); },
      };
    });
  });
})();

File: pipeline/modelspace.py
"""Model space: the same planet with one physical assumption changed.

`assumed_cloud_state`, `assumed_metallicity` and the planet's orbital distance are inputs the
pipeline is *forced* to pick. Elsewhere on the site they are a disclaimer. Here they become
axes you can move, so a reader can see for themselves how much of a swatch is measurement and
how much is the model's opinion. Three views, one mechanism:

- **migration track** — colour as a function of orbital distance. Methane freezes out, clouds
  condense, the colour walks across the spectrum. This is the site's thesis made draggable.
- **what-if variants** — the same planet with the clouds stripped, the cloud deck closed, or
  the metallicity forced to the ends of its plausible range.
- **colour year** — for an eccentric orbit, the distance track resampled at r(t) over one
  period. Not a separate model: literally the migration track, indexed by time.

THE ANCHORING RULE (why every one of these agrees with the planet's published colour).
Each view is computed as the planet's OWN albedo multiplied by a model RESPONSE RATIO:

    A_view(λ) = A_planet(λ) × [ A_model(λ | changed) / A_model(λ | as published) ]

At the home position the ratio is exactly 1, so the slider starts on the planet's real colour
no matter which engine produced it — parametric, Cahoy grid, or a measured spectrum. The model
contributes only how the colour MOVES, never the identity it moves from. This is the same
idiom `pipeline.spectrum.phase` already uses for its "cahoy-ratio" tier, for the same reason.

WHICH ENGINE DRIVES THE RESPONSE. The parametric model (pipeline.spectrum.parametric), always.
It is the only engine that is continuous and defined over the whole range these features need:
real orbits run from 0.008 AU to 240 AU, and an eccentric planet like HD 80606 b (e = 0.93)
sweeps 0.03 -> 0.89 AU within a single orbit. The Cahoy grid covers 0.8-10 AU only, so a
grid-driven track would flatline against its clamp across most of that motion, and switching
engines mid-track would put a seam in the middle of the slider that reads as a bug.

The grid is not wasted: `cahoy_reference_points` returns what Cahoy et al. 2010 actually says
at its own four distances, for the planets it covers, so the page can mark the reference grid
the Roman Coronagraph community uses right on the slider and let the two be compared.

HONESTY LIMITS, which the UI must carry:
- Equilibrium temperature is scaled as T ∝ r^(-1/2) from the planet's own published T_eq. That
  holds the Bond albedo and the star fixed, which is the standard first-order treatment and
  exactly reproduces the published temperature at the home position.
- Every stop is instantaneous radiative equilibrium. Real atmospheres have thermal inertia and
  lag; cloud decks and chemistry lag much more. The colour year is "where equilibrium points",
  not a forecast of what a telescope would see on a given night.
- Nothing here is a measurement. Moving a planet is a thought experiment about a model.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, replace

import numpy as np
from colour import XYZ_to_Lab, delta_E

from pipeline.colour.cie import ColourResult, reflected_flux_to_colour
from pipeline.config import GRID_NM
from pipeline.spectrum.base import ProviderUnavailable
from pipeline.spectrum.cahoy_grid import CAHOY_DISTANCES_AU, grid_albedo_at
from pipeline.spectrum.parametric import model_for

# --- The distance axis -------------------------------------------------------------------
# The slider is laid out in TEMPERATURE, not in orbital distance, and this matters.
#
# A range fixed in distance (say 1/32 to 32 times the real orbit) makes the feature dead for
# exactly the planets it should be best on: Neptune's own orbit is so cold that five octaves
# inward still only reaches 266 K, so its colour sits still for the whole slider. All the
# interesting chemistry — methane freezing out, water clouds condensing, alkali metals turning
# up — happens between roughly 100 K and 2000 K, and every planet should cross it.
#
# So the endpoints are temperatures, converted to distance per planet via r ∝ T^-2. The home
# position lands wherever the planet's real orbit falls in that range, which is itself worth
# seeing: it shows where this world sits among all the regimes it could have occupied.
_TEQ_LO_K = 40.0  # colder than Neptune: everything condensed, nothing left to change
_TEQ_HI_K = 2600.0  # hotter than KELT-9 b: cloud-free, alkali-soaked, very dark
# Always give the slider at least two octaves of travel either way, so planets already sitting
# at an extreme of that temperature range still have something to drag.
_MIN_RATIO_CEILING = 0.25
_MIN_RATIO_FLOOR = 4.0
N_STOPS = 25


def distance_ratios(eq_temp_k: float) -> tuple[tuple[float, ...], int]:
    """The slider's stops for one planet, as multiples of its real orbital distance, plus the
    index of the home position.

    Log-spaced across the distance range that spans `_TEQ_LO_K`-`_TEQ_HI_K`, with the exact
    home ratio 1.0 inserted so the slider has a stop that reproduces the published colour
    exactly rather than one that merely passes near it.
    """
    temp = max(eq_temp_k, 1.0)
    lo = min((temp / _TEQ_HI_K) ** 2, _MIN_RATIO_CEILING)
    hi = max((temp / _TEQ_LO_K) ** 2, _MIN_RATIO_FLOOR)
    spaced = [
        math.exp(math.log(lo) + (math.log(hi) - math.log(lo)) * i / (N_STOPS - 2))
        for i in range(N_STOPS - 1)
    ]
    home = sum(1 for r in spaced if r < 1.0)
    return tuple(spaced[:home] + [1.0] + spaced[home:]), home

# Samples around one orbit for the colour year, in EQUAL STEPS OF TIME (not of true anomaly).
# That distinction is the whole character of an eccentric orbit: the planet whips through
# periastron and loiters at apoastron, so the colour should sit cool for most of the loop and
# flash hot briefly. Sampling the angle uniformly would show the opposite and be a lie.
COLOUR_YEAR_SAMPLES = 48
# Below this eccentricity the loop is visually a still frame; not worth a control on the page.
COLOUR_YEAR_MIN_ECCENTRICITY = 0.1


@dataclass(frozen=True)
class TrackStop:
    """One position of the migration slider."""

    r_over_a: float  # orbital distance as a multiple of the planet's real one
    au: float | None  # absolute distance, when the real semi-major axis is known
    eq_temp_k: float  # equilibrium temperature there (T ∝ r^-1/2 from the published value)
    hex: str
    luminance_y: float  # true relative brightness, so "dark" survives the trip


@dataclass(frozen=True)
class CahoyPoint:
    """What the Cahoy et al. 2010 grid itself says at one of its four native distances,
    for planets it covers. The reference mark against the modelled track."""

    au: float
    r_over_a: float
    hex: str
    in_track_range: bool  # whether this distance falls within the slider's span


@dataclass(frozen=True)
class MigrationTrack:
    home_index: int
    stops: tuple[TrackStop, ...]
    # Real Cahoy grid colours at 0.8/2/5/10 AU, empty when the planet is outside what the
    # grid covers (rocky worlds, or no grid installed).
    cahoy_points: tuple[CahoyPoint, ...] = ()


@dataclass(frozen=True)
class WhatIfVariant:
    """The planet with exactly one modelling assumption changed."""

    id: str
    label: str  # plain English, self-explanatory without the site's vocabulary
    detail: str  # what was changed and what it means physically
    hex: str
    luminance_y: float
    delta_e2000: float  # perceptual distance from the planet's published colour


@dataclass(frozen=True)
class ColourYear:
    """One orbit of an eccentric planet, sampled at equal steps of time."""

    eccentricity: float
    period_days: float | None
    periastron_au: float | None
    apoastron_au: float | None
    # Index into MigrationTrack.stops (fractional — the client interpolates) for each sample.
    track_positions: tuple[float, ...]
    # Fraction of the orbit spent within one octave of periastron: the "flash" duty cycle.
    hot_fraction: float


# --- Response ratios ---------------------------------------------------------------------


def _safe_ratio(new: np.ndarray, ref: np.ndarray) -> np.ndarray:
    """A_new / A_ref, guarded where the reference goes ~black (deep absorption bands), which
    would otherwise blow a spike through the ratio. Those channels take the spectral-mean
    ratio instead — the same guard `pipeline.spectrum.phase` uses, for the same reason."""
    valid = ref > 0.01
    if not np.any(valid):
        return np.ones_like(ref)
    mean_ratio = float(np.mean(new[valid] / ref[valid]))
    return np.where(valid, new / np.maximum(ref, 1e-9), mean_ratio)


def _param_albedo(
    *,
    eq_temp_k: float,
    radius_r_earth: float | None,
    mass_m_earth: float | None,
    metallicity: float | None = None,
    cloud_fraction: float | None = None,
) -> np.ndarray:
    """The parametric model's albedo under a given set of knobs, on GRID_NM."""
    model = model_for(
        equilibrium_temp_k=eq_temp_k,
        radius_r_earth=radius_r_earth,
        mass_m_earth=mass_m_earth,
        metallicity_override=metallicity,
    )
    albedo = model.albedo
    if cloud_fraction is not None:
        albedo = replace(albedo, cloud_fraction=cloud_fraction)
    return albedo.geometric_albedo(GRID_NM)


def eq_temp_at(eq_temp_k: float, r_over_a: float) -> float:
    """Equilibrium temperature at a multiple of the planet's real orbital distance.

    T_eq ∝ r^(-1/2) at fixed Bond albedo and stellar luminosity. Scaling the planet's OWN
    published T_eq (rather than recomputing it from stellar radius and Teff) is deliberate: it
    reproduces the published number exactly at r_over_a = 1, so the slider's home position can
    never disagree with the temperature printed elsewhere on the page.
    """
    return eq_temp_k / math.sqrt(max(r_over_a, 1e-9))


# --- The three views ---------------------------------------------------------------------


def _colour(albedo: np.ndarray, star_flux: np.ndarray) -> ColourResult:
    return reflected_flux_to_colour(
        np.clip(albedo, 0.0, 1.0) * star_flux,
        method="full-spectrum",
        illuminant_flux=star_flux,
        confidence="low",  # a thought experiment about a model, never a measurement
    )


def migration_track(
    *,
    base_albedo: np.ndarray,
    star_flux: np.ndarray,
    eq_temp_k: float,
    semi_major_axis_au: float | None,
    radius_r_earth: float | None,
    mass_m_earth: float | None,
    metallicity: float | None = None,
) -> MigrationTrack:
    """Colour as the planet is dragged in and out of its orbit.

    `base_albedo` is the planet's published albedo spectrum, whatever engine produced it; the
    parametric model supplies only the response ratio (see the module docstring).
    """
    ref = _param_albedo(
        eq_temp_k=eq_temp_k, radius_r_earth=radius_r_earth, mass_m_earth=mass_m_earth
    )
    base = np.clip(np.asarray(base_albedo, dtype=float), 0.0, 1.0)
    ratios, home_index = distance_ratios(eq_temp_k)

    stops: list[TrackStop] = []
    for ratio in ratios:
        temp = eq_temp_at(eq_temp_k, ratio)
        moved = _param_albedo(
            eq_temp_k=temp, radius_r_earth=radius_r_earth, mass_m_earth=mass_m_earth
        )
        colour = _colour(base * _safe_ratio(moved, ref), star_flux)
        stops.append(
            TrackStop(
                r_over_a=round(ratio, 5),
                au=round(semi_major_axis_au * ratio, 5) if semi_major_axis_au else None,
                eq_temp_k=round(temp, 1),
                hex=colour.hex,
                luminance_y=round(colour.luminance_y, 5),
            )
        )

    return MigrationTrack(
        home_index=home_index,
        stops=tuple(stops),
        cahoy_points=cahoy_reference_points(
            star_flux=star_flux,
            semi_major_axis_au=semi_major_axis_au,
            radius_r_earth=radius_r_earth,
            metallicity=metallicity,
            ratio_range=(ratios[0], ratios[-1]),
        ),
    )


def cahoy_reference_points(
    *,
    star_flux: np.ndarray,
    semi_major_axis_au: float | None,
    radius_r_earth: float | None,
    metallicity: float | None,
    ratio_range: tuple[float, float],
) -> tuple[CahoyPoint, ...]:
    """The Cahoy et al. 2010 grid's own answer at its four native distances, lit by this
    planet's star. Reference marks for the slider, so the modelled track can be read against
    the published grid rather than just asserted.

    Empty for rocky worlds (the grid is Jupiter/Neptune-class only), for planets with no known
    orbit, and whenever the grid is not installed — never a broken build.

    Note the grid is computed for a Sun-like host; its distances are used as published rather
    than rescaled by this star's luminosity, matching how `pipeline.spectrum.cahoy_grid` picks
    a grid point for the planet's own colour. The page says so.
    """
    if semi_major_axis_au is None or metallicity is None:
        return ()
    if radius_r_earth is not None and radius_r_earth < 2.0:
        return ()  # rocky: neither Cahoy class applies

    lo, hi = ratio_range
    points: list[CahoyPoint] = []
    for au in CAHOY_DISTANCES_AU:
        try:
            albedo = grid_albedo_at(GRID_NM, dist_au=au, metallicity=metallicity)
        except ProviderUnavailable:
            return ()
        ratio = au / semi_major_axis_au
        points.append(
            CahoyPoint(
                au=au,
                r_over_a=round(ratio, 5),
                hex=_colour(albedo, star_flux).hex,
                in_track_range=lo <= ratio <= hi,
            )
        )
    return tuple(points)


def what_if_variants(
    *,
    base_albedo: np.ndarray,
    star_flux: np.ndarray,
    base_colour_xyz: tuple[float, float, float],
    eq_temp_k: float,
    radius_r_earth: float | None,
    mass_m_earth: float | None,
    metallicity: float | None,
) -> tuple[WhatIfVariant, ...]:
    """The same planet under a handful of alternative modelling assumptions.

    Cloud variants apply to everything (even a rocky world has a cloud fraction). Metallicity
    variants are skipped where `metallicity` is None — rocky worlds have no H/He envelope, so
    the giant mass-metallicity relation is meaningless for them and the pipeline stores None.
    """
    ref = _param_albedo(
        eq_temp_k=eq_temp_k, radius_r_earth=radius_r_earth, mass_m_earth=mass_m_earth
    )
    base = np.clip(np.asarray(base_albedo, dtype=float), 0.0, 1.0)
    base_lab = XYZ_to_Lab(np.array(base_colour_xyz))

    specs: list[tuple[str, str, str, dict]] = [
        (
            "clouds-off",
            "No clouds at all",
            "Strip the cloud deck and you see straight down into the deep atmosphere: much "
            "darker, and bluer where Rayleigh scattering takes over.",
            {"cloud_fraction": 0.0},
        ),
        (
            "clouds-thick",
            "Solid cloud deck",
            "Close the cloud deck completely. Clouds are bright and fairly grey, so this "
            "pushes almost any planet toward a pale version of its star's own colour.",
            {"cloud_fraction": 1.0},
        ),
    ]
    if metallicity is not None:
        specs += [
            (
                "metal-poor",
                "Solar metallicity (1×)",
                "As few heavy elements as the Sun. Less methane to absorb the red end, so "
                "the colour drifts back toward the star's.",
                {"metallicity": 1.0},
            ),
            (
                "metal-rich",
                "Metal-rich (30×)",
                "Thirty times the Sun's heavy elements — Neptune territory. Deeper methane "
                "bands eat the red and the planet swings blue-green.",
                {"metallicity": 30.0},
            ),
        ]

    variants: list[WhatIfVariant] = []
    for vid, label, detail, knobs in specs:
        changed = _param_albedo(
            eq_temp_k=eq_temp_k,
            radius_r_earth=radius_r_earth,
            mass_m_earth=mass_m_earth,
            **knobs,
        )
        colour = _colour(base * _safe_ratio(changed, ref), star_flux)
        variants.append(
            WhatIfVariant(
                id=vid,
                label=label,
                detail=detail,
                hex=colour.hex,
                luminance_y=round(colour.luminance_y, 5),
                delta_e2000=round(
                    float(delta_E(base_lab, XYZ_to_Lab(np.array(colour.xyz)), method="CIE 2000")),
                    1,
                ),
            )
        )
    return tuple(variants)


# --- Colour year -------------------------------------------------------------------------


def _true_anomaly(mean_anomaly: float, e: float) -> float:
    """Solve Kepler's equation M = E - e sin E by Newton's method, then convert the eccentric
    anomaly E to the true anomaly ν. Converges in a handful of iterations for every archive
    eccentricity; the starting guess is the standard one that stays stable near e -> 1."""
    m = mean_anomaly
    ecc = min(max(e, 0.0), 0.999)
    guess = m if ecc < 0.8 else math.pi
    for _ in range(60):
        f = guess - ecc * math.sin(guess) - m
        fp = 1.0 - ecc * math.cos(guess)
        step = f / fp
        guess -= step
        if abs(step) < 1e-12:
            break
    return 2.0 * math.atan2(
        math.sqrt(1.0 + ecc) * math.sin(guess / 2.0),
        math.sqrt(1.0 - ecc) * math.cos(guess / 2.0),
    )


def colour_year(
    *,
    track: MigrationTrack,
    eccentricity: float | None,
    semi_major_axis_au: float | None,
    period_days: float | None = None,
    samples: int = COLOUR_YEAR_SAMPLES,
) -> ColourYear | None:
    """Where on the migration track the planet sits at each moment of one orbit.

    Returns None for orbits too circular to be worth animating. The output is positions on the
    track, not colours: the colours are the track's, which is what guarantees the loop and the
    slider can never disagree.
    """
    if eccentricity is None or eccentricity < COLOUR_YEAR_MIN_ECCENTRICITY:
        return None

    ecc = min(eccentricity, 0.999)
    log_ratios = [math.log(s.r_over_a) for s in track.stops]

    positions: list[float] = []
    near = 0
    for i in range(samples):
        nu = _true_anomaly(2.0 * math.pi * i / samples, ecc)
        r_over_a = (1.0 - ecc**2) / (1.0 + ecc * math.cos(nu))
        # Fractional index into the track. The stops are log-spaced but not perfectly evenly
        # (the exact home ratio is inserted among them), so interpolate on the real values.
        positions.append(
            round(float(np.interp(math.log(r_over_a), log_ratios, range(len(log_ratios)))), 3)
        )
        if r_over_a <= 2.0 * (1.0 - ecc):  # within one octave of periastron
            near += 1

    return ColourYear(
        eccentricity=round(eccentricity, 4),
        period_days=period_days,
        periastron_au=(
            round(semi_major_axis_au * (1.0 - ecc), 5) if semi_major_axis_au else None
        ),
        apoastron_au=(
            round(semi_major_axis_au * (1.0 + ecc), 5) if semi_major_axis_au else None
        ),
        track_positions=tuple(positions),
        hot_fraction=round(near / samples, 3),
    )
